### PR TITLE
[linux/qemu] Qemu bridging v2

### PR DESCRIPTION
**Pull request by sharder996**
Public side of [#647](https://github.com/canonical/multipass-private/pull/647)

The spec for this feature can be found [here](https://docs.google.com/document/d/1UAZgCHHjdY_RmYCIir-p9sDpa2M7uRP7P81Tx2qM0c8/edit).

This second version of QEMU network bridging was required because of the change in implementation of `find_bridge_with()`. Since most of bridged networks was already available on macOS/QEMU and implemented, this PR is mostly just for enabling and wiring up the function calls needed to create a network interface on Linux.

The key difference on Linux is the presence of `bridge_helper.c`. Below is an excerpt from the spec on the purpose of the bridge helper:

> Finally, to use bridges on QEMU, on Linux, we will use a bridge helper. The QEMU bridge helper,  shipped with QEMU, is a simple program which checks a file in `/etc/` to see if using a certain bridge is allowed; then it sets up the devices needed for QEMU to use the bridge. Access to this file would be denied in our snap. Instead, our approach will be to fork the source of the bridge helper and remove the need to check for this file, thus allowing all the bridges we create. Then, from the QEMU command line, we must specify the path to our binary, for avoiding executing the default bridge helper.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -240,7 +240,9 @@ else()
       add_definitions(-DMULTIPASS_COMPILER_GCC)
     endif()
 
-    add_compile_options(-Wextra -Wempty-body -Wformat-security -Winit-self -Warray-bounds -Wnon-virtual-dtor)
+    add_compile_options(-Wextra -Wempty-body -Wformat-security -Winit-self -Warray-bounds)
```
Comment (ricab) [inline CMakeLists.txt:243]: Why is this needed?

File: src/platform/backends/qemu/linux/qemu_platform_detail_linux.cpp [hunk 2]
```diff
@@ -192,27 +195,47 @@ QStringList mp::QemuPlatformDetail::vm_platform_args(const VirtualMachineDescrip
...
 }
 
-mp::QemuPlatform::UPtr mp::QemuPlatformFactory::make_qemu_platform(const Path& data_dir) const
+void mp::QemuPlatformDetail::set_authorization(std::vector<NetworkInterfaceInfo>& networks)
 {
-    return std::make_unique<mp::QemuPlatformDetail>(data_dir);
+    const auto& br_nomenclature = MP_PLATFORM.bridge_nomenclature();
+
+    for (auto& net : networks)
+        if (net.type == "ethernet" && net.needs_authorization &&
```
Comment (ricab) [inline src/platform/backends/qemu/linux/qemu_platform_detail_linux.cpp:232]: Tiny irrelevant detail: better place the check for `net.needs_authorization` first and avoid the string comparison in many cases.

File: src/platform/backends/qemu/linux/qemu_platform_detail_linux.cpp [hunk 3]
```diff
@@ -192,27 +195,47 @@ QStringList mp::QemuPlatformDetail::vm_platform_args(const VirtualMachineDescrip
...
+            mpu::find_bridge_with(networks, net.id, br_nomenclature))
+            net.needs_authorization = false;
```
Comment (ricab) [inline src/platform/backends/qemu/linux/qemu_platform_detail_linux.cpp:234]: But where is this ever set to true? 

In LXD, it is set in [here](https://github.com/canonical/multipass/blob/1f67f00527092c7a571529a86986fed4ae7d7add/src/platform/backends/lxd/lxd_virtual_machine_factory.cpp#L64-L67). Seeing as this is really a platform-level thing, I am thinking that should instead be set immediately when reading networks from the system in the Platform. WDYT?

**Review by ricab — COMMENTED** [review]
Hey @sharder996, thank you for your work on this. I did a quick pass and I have a couple of questions.

File: CMakeLists.txt (hunk 1 above)
Comment (sharder996) [reply CMakeLists.txt:243]: The `non-virtual-dtor` is not a valid flag when compiling the `bridge_helper.c` which is why it's explicitly added to `CMAKE_CXX_FLAGS` a few lines below.

File: src/platform/backends/qemu/linux/qemu_platform_detail_linux.cpp (hunk 3 above)
Comment (sharder996) [reply src/platform/backends/qemu/linux/qemu_platform_detail_linux.cpp:234]: Like setting it [here](https://github.com/canonical/multipass/blob/1f67f00527092c7a571529a86986fed4ae7d7add/src/platform/platform_linux.cpp#L152), then?

Comment (ricab) [reply src/platform/backends/qemu/linux/qemu_platform_detail_linux.cpp:234]: That is what I was thinking, but now I am not sure anymore, because not all backends need that. Namely Virtualbox, which can just connect to an interface without a dedicated switch. I guess the easiest for now is to just add the logic to necessary in here — but set it to `true` as necessary, not to `false` (unless I am missing something).

File: CMakeLists.txt (hunk 1 above)
Comment (ricab) [reply CMakeLists.txt:243]: Hmm, couldn't we just compile the bridge helper as cpp? It was already adapted for us, right?

Comment (ricab) [reply CMakeLists.txt:243]: Meh, nevermind. Leave it

**Comment by georgeliao** [comment]
@sharder996 
Since I was not in the discussions, it would be nice to have a context introduction. Something like why bridge_helper executable is needed on linux but not on macos. What was the state of the code before and why did you do some refactoring?

**Comment by sharder996** [comment]
Does clang-format format `test_qemu_linux_platform.cpp` the same as the linter in CI for anyone else? Don't know why is doesn't format the same for me.

**Comment by georgeliao** [comment]
@sharder996 
Check the clang-format executable version, check the clang-format configuration file your local IDE uses is the in-repo one by changing the file and see if re-apply formatting makes a difference.

**Review by georgeliao — CHANGES_REQUESTED** [review]
@sharder996 
It looks very good. I only have a few minor comments. The unit test coverage is not high enough, it is easy to add these tests as well. However, on the other side, the uncovered part does not have any actual logic, so it is ok to skip them. I will leave it to you whether you want to push the test coverage up.  

Funcitonal testing was done as well, it works fine on linux and macos.

**Comment by sharder996** [comment]
@georgeliao As for the missing test line coverage. Having no logic to actually test was the reason why I didn't pursue it.

File: tests/qemu/test_qemu_backend.cpp [hunk 4]
```diff
@@ -801,15 +802,24 @@ TEST_F(QemuBackend, createsQemuSnapshotsFromJsonFile)
...
+    auto [mock_platform, guard] = mpt::MockPlatform::inject();
+    EXPECT_CALL(*mock_platform, get_network_interfaces_info)
+        .WillOnce(Return(std::map<std::string, mp::NetworkInterfaceInfo>{
+            {"lxdbr0", {"lxdbr0", "bridge", "gobbledygook"}},
+            {"mpbr0", {"mpbr0", "bridge", "gobbledygook"}},
+            {"virbr0", {"virbr0", "bridge", "gobbledygook"}},
+            {"mpqemubr0", {"mpqemubr0", "bridge", "gobbledygook"}},
+            {"enxe4b97a832426", {"enxe4b97a832426", "ethernet", "gobbledygook"}}}));
+
+    EXPECT_NO_THROW(backend.networks());
```
[the remark this answers is not in the document]
Comment (georgeliao) [reply tests/qemu/test_qemu_backend.cpp]: I do not think what you are trying to achieve is possible. 

The syntax `mock_qemu_platform->QemuPlatform::is_network_supported(arg)` is using the derived class (`MockQemuPlatform`) instance to call the base class (`QemuPlatform`) function implementation because of function hiding. This is useful when you want to call the base class function from a derived class instance. However, this only works if the base class has the function implementation and it is not the case here because `QemuPlatform::is_network_supported` is pure virtual. 

And there is no way it can somehow dispatch to `QemuPlatformDetail::is_network_supported`, it is a different derived class of `QemuPlatform` and `MockQemuPlatform` knows nothing about it, he only knows himself and the base class. 

Feel free to play with the [code snippet](https://godbolt.org/z/cd7qYnT14) to understand better.

[the remark this answers is not in the document]
Comment (sharder996) [reply tests/qemu/test_qemu_backend.cpp]: Yeah, I kinda thought as much. I wanted to do something like this in order to get some meaningful coverage on some of the new functions in the qemu platform.

File: src/platform/backends/qemu/linux/qemu_platform_detail_linux.cpp [hunk 5]
```diff
@@ -192,27 +195,46 @@ QStringList mp::QemuPlatformDetail::vm_platform_args(const VirtualMachineDescrip
          << QString::fromStdString(fmt::format("tap,ifname={},script=no,downscript=no,model=virtio-net-pci,mac={}",
                                                tap_device_name, vm_desc.default_mac_address));
 
+    auto bridge_helper_path = QCoreApplication::applicationDirPath() + "/bridge_helper";
```
[the remark this answers is not in the document]
Comment (ricab) [reply src/platform/backends/qemu/linux/qemu_platform_detail_linux.cpp]: If it is not difficult to do, I would say yes. If you start running into trouble and needing to invest more time for any reason, I would say leave it.

**Review by georgeliao — DISMISSED** [review]

**Review by ricab — APPROVED** [review]
Good stuff, well done!
